src/server/frontend_wayland/wl_seat.h

```diff
@@ -60,6 +63,14 @@ class WlSeat : public wayland::Seat::Global
 
     static auto from(struct wl_resource* seat) -> WlSeat*;
 
+    /// Returns the timestamp of the event in milliseconds, suitable for Wayland events
+    /// @{
+    static auto timestamp_of(MirInputEvent const* event) -> uint32_t;
+    static auto timestamp_of(MirKeyboardEvent const* event) -> uint32_t;
+    static auto timestamp_of(MirPointerEvent const* event) -> uint32_t;
+    static auto timestamp_of(MirTouchEvent const* event) -> uint32_t;
+    /// @}
+
```

Comment:
That's a set `static` methods with no obvious connection to `WlSeat`.

1. shouldn't they be at namespace scope?
2. is this the right header? `wl_timestamp.h`?

Reply:
Should have explained this better, 2 reasons:
1. In the next phase the seat will manage mapping events to serials, which means it will deal with events and everything that uses events will already be using it
2. The timestamp conversion logic is a little messy and I'd rather not scatter it around the code base else it might get out of sync

I supposed I could make `timestamp_of(MirInputEvent)` on the seat and then `namespace {}` functions for calling it with the various specific types, or I could make `timestamp_of()` a standalone function in `wayland_utils.h` or something.

Reply:
I'm not sure I follow your response.

1. I still don't see why pure functions that take a `Mir*Event` and return a timestamp are members of `WlSeat`, nor why they are declared in the header that defines it. They are perfectly usable without any of `WlSeat`.
2. Why would you think I'm concerned about reducing the number of overloads? They all have the same semantics (I would be worried if they didn't).

Reply:
The issue is that I need *somewhere* to put it, though I suppose that isn't the best argument. Is making them standalone functions declared and implemented in the seat files better? Where do you think they should go?

Reply:
Perhaps the correct place is the event system.

Reply:
OK